### ISSUE TITLE
Pretty Low IoU@0.7 While Using PnPDA

### ISSUE THREAD

**Issue by Xhy-5000** [luotianyou349/PnPDA#3]

**Comment by luotianyou349** [comment]
Thanks for your question. I retested my model today and got the following results:
![Image](https://github.com/user-attachments/assets/a18748a0-183e-4ea4-93ed-505e577bfc8e)
I think you didn't get a good model in the pre-training stage or a good encoder, which led to poor final results, because your config file does not seem to have any problems. You can check the detection accuracy of the encoder and the semantic similarity of the positive and negative sample pairs in the pre-training stage.

**Comment by Xhy-5000** [comment]
Thank you for the instructions. I think my encoders have pretty good performance, their IoU@0.7 is over 0.8. But I'm not quite sure how to enable the model quality during the pre-training stage. Do you have any suggestions? Here is my configuration file for the pre-training stage. 
```
name: pp8_pp4_ssl_v2xset
root_dir: '/mnt/ct/v2xbackdoor/v2x-vit/v2xset/train'
validate_dir: '/mnt/ct/v2xbackdoor/v2x-vit/v2xset/test'
# root_dir: '/mnt/ct/v2xbackdoor/v2x-vit/opv2v/train'
# validate_dir: '/mnt/ct/v2xbackdoor/v2x-vit/opv2v/test'
model_dir: 'model/'

yaml_parser: "load_pp_pp_params"
train_params:
  batch_size: &batch_size 2
  epoches: 30
  eval_freq: 1
  save_freq: 1
  max_cav: &max_cav 5

fusion:
  core_method: 'IntermediateFusionDatasetV3' # LateFusionDataset, EarlyFusionDataset, IntermediateFusionDataset supported
  args: 
    # same_data: false
    # full_data: false
    # flip_data: true
    # is_augment: false
    same_data: false
    full_data: true
    flip_data: false
    is_augment: false

# preprocess-related
preprocess_q:
  # options: BasePreprocessor, VoxelPreprocessor, BevPreprocessor
  core_method: 'SpVoxelPreprocessor'
  args:
    voxel_size: &voxel_size_q [0.8, 0.8, 4]
    max_points_per_voxel: 32
    max_voxel_train: 32000
    max_voxel_test: 70000
  # lidar range for each individual cav.
  cav_lidar_range: &cav_lidar [-140.8, -40, -3, 140.8, 40, 1]
  
# preprocess-related
preprocess_k:
  # options: BasePreprocessor, VoxelPreprocessor, BevPreprocessor
  core_method: 'SpVoxelPreprocessor'
  args:
    voxel_size: &voxel_size_k [0.4, 0.4, 4]
    max_points_per_voxel: &T 32
    max_voxel_train: 36000
    max_voxel_test: 70000
  # lidar range for each individual cav.
  cav_lidar_range: [-140.8, -40, -3, 140.8, 40, 1]

data_augment:
  - NAME: random_world_flip
    ALONG_AXIS_LIST: [ 'x' ]

  - NAME: random_world_rotation
    WORLD_ROT_ANGLE: [ -0.78539816, 0.78539816 ]

  - NAME: random_world_scaling
    WORLD_SCALE_RANGE: [ 0.95, 1.05 ]

# anchor box related
postprocess:
  core_method: 'VoxelPostprocessor' # VoxelPostprocessor, BevPostprocessor supported
  anchor_args:
    cav_lidar_range: *cav_lidar
    l: 3.9
    w: 1.6
    h: 1.56
    r: [0, 90]
    feature_stride: 2
    num: &achor_num 2
  target_args:
    pos_threshold: 0.6
    neg_threshold: 0.45
    score_threshold: 0.20
  voxel_size: *voxel_size_q
  cav_lidar_range: *cav_lidar
  order: 'hwl' # hwl or lwh
  max_num: 100 # maximum number of objects in a single frame. use this number to make sure different frames has the same dimension in the same batch
  nms_thresh: 0.15

# model related
model:
  core_method: ssl
  args:
    encoder_q:
      core_method: point_pillar_encoder
      saved_pth: /mnt/xhy/MPDA/opencood/logs/pp8_v2xset/net_epoch20.pth
      # saved_pth: /mnt/xhy/MPDA/opencood/logs/pp8_opv2v/net_epoch20.pth
      args:
        voxel_size: *voxel_size_q
        lidar_range: *cav_lidar
        anchor_number: *achor_num
        max_cav: *max_cav
        channel: &c_q 256
        featrue_size: &s_q [50,176]

        pillar_vfe:
          use_norm: true
          with_distance: false
          use_absolute_xyz: true
          num_filters: [64]
          
        point_pillar_scatter:
          num_features: 64

        base_bev_backbone:
          layer_nums: [3, 5]
          layer_strides: [2, 2]
          num_filters: [64, 128]
          upsample_strides: [1, 2]
          num_upsample_filter: [128, 128]

    encoder_k:
      core_method: point_pillar_encoder
      saved_pth: /mnt/xhy/MPDA/opencood/logs/pp4_v2xset/net_epoch20.pth
      # saved_pth: /mnt/xhy/MPDA/opencood/logs/pp4_opv2v/net_epoch20.pth
      args:
        voxel_size: *voxel_size_k
        lidar_range: *cav_lidar
        anchor_number: *achor_num
        max_cav: *max_cav
        channel: &c_k 384
        featrue_size: &s_k [100,352]
  
        pillar_vfe:
          use_norm: true
          with_distance: false
          use_absolute_xyz: true
          num_filters: [64]
          
        point_pillar_scatter:
          num_features: 64
  
        base_bev_backbone:
          layer_nums: [3, 5, 8]
          layer_strides: [2, 2, 2]
          num_filters: [64, 128, 256]
          upsample_strides: [1, 2, 4]
          num_upsample_filter: [128, 128, 128]
    
    compressor:
      in_channel: *c_k
      in_size: *s_k
      featrue_size: *s_q
      channel: *c_q
    
    projector:
      num_layers: 1 #2
      num_heads: 16
      d_ff: 512
      d_model: *c_q
      dropout: 0.1

    predictor:
      num_layers: 1
      num_heads: 16
      d_model: *c_q
      d_ff: 256
      
    anchor_num: *achor_num
    momentum: 0.6
    

loss:
  core_method: contrastive_learning_loss
  args:
    tau: 0.1
    max_voxel: 40

optimizer:
  core_method: Adam
  lr: 0.001
  args:
    eps: 1e-8
    weight_decay: 1e-4

lr_scheduler:
  core_method: multistep #step, multistep and Exponential support
  gamma: 0.1
  step_size: [10, 50, 80]


```

**Comment by luotianyou349** [comment]
You can get a better model by adjusting tau and lr, which can be judged by observing the semantic similarity of positive sample pairs. Try it a few more times, good luck!

**Comment by Xhy-5000** [comment]
Thank you very much, I'm trying to do so.
